CRAYSAT-1487: Add documentation to set up external SAT (#12)
Cray-HPE/docs-sat: docs/install.md

    the modified CFS configuration. If more than one CFS configuration was
    modified, use the first one.

## Optional: Configuring SAT to run on an external system over the CAN

> It is possible to configure SAT so that it can interact with CSM over the CAN. This allows SAT to run from a physical or
> virtual machine which is not a part of the CSM management cluster.

Reviewer: If we add steps on installing SAT in a python virtualenv, we can say something like this (includes minor rewording/shortening):

```suggestion
SAT can optionally be installed and configured on an external system to interact with CSM over the CAN.
```

### Limitations

> Some SAT subcommands must interact with services which run on the NCNs in the CSM management cluster, including CSM REST
> APIs, Kubernetes, Ceph, and S3. Certain SAT commands, such as `sat bootsys`, may not work as expected.

Reviewer: I'm not exactly sure what this is trying to say. I think you're trying to convey that some SAT functionality which uses externally accessible APIs will work, while other functionality which depends on functionality local to the management NCNs will not. Here's my suggestion:

```suggestion
Most SAT subcommands work by accessing APIs which are reachable via the CAN. However, certain SAT commands depend on
host-based functionality on the management NCNs and will not work from an external system. This includes the following:

- The `platform-services` and `ncn-power` stages of `sat bootsys`
- The local host information displayed by the `--local` option of `sat showrev`
```

Author: I think this is an improvement, but since I didn't do a comprehensive test of all of the SAT functionality, I didn't want an enumerated list to imply that it was exclusive, i.e. that everything else should be expected to work. I suppose a compromise could be to specify `this includes, but may not necessarily be limited to the following:`?

> Instructions for installing SAT on an external system are provided "as-is", and installing SAT on an external system is
> not an officially supported configuration. These instructions are provided with the hope that they can useful for users
> who desire additional flexibility.

Reviewer: I think this can be shortened a bit:

```suggestion
Installing SAT on an external system is not an officially supported configuration. These instructions are provided
"as-is" with the hope that they can useful for users who desire additional flexibility.
```

Author: I like this, it seems clearer and more concise.

Certain additional steps may need to be taken to install and configure SAT depending on the configuration of the
external system in use. These additional steps may include provisioning virtual machines, installing packages, or
configuring TLS certificates, and these steps are outside the scope of this documentation. This section covers only the
steps needed to configure SAT to use externally-accessible API endpoints exposed by CSM.

### Prerequisites

- The external system must be on the Customer Access Network (CAN).

> - The SAT Python package and its dependencies, or the SAT podman image and wrapper script, have been installed on the
>   external system.

Reviewer: If we document installing sat into a python virtualenv:

```suggestion
- Python 3 is installed on the external system.
```

Author: I would take this a step further and specify a minimum of Python 3.7, which is I believe the minimum required by some of our dependencies. FWIW I was testing my virtualenv with Python 3.9 so I might run it again with 3.7.

Reviewer: Ahh, yes, that's right.

- `kubectl` is installed on the external system.
- The root CA certificates used when installing CSM have been added to the external system's trust store such that
  authenticated TLS connections can be made to the CSM REST API gateway. See the [Certificate
  Authority](https://cray-hpe.github.io/docs-csm/en-12/background/certificate_authority/) section of the CSM
  documentation for more information.

> ### Procedure

Reviewer: I think it's a good idea to document installing from a virtualenv because of all the small difficulties that add up when using sat in a container.

To do that, I'd suggest that we add steps at the beginning to clone the sat repo, create a Python 3 virtual env, activate it, install our locked requirements, and then install sat. I suggest locked requirements because there is then less likelihood of something going wrong here that is unique to the deps installed in their environment.

I'm open to discussion on this though.

> 1. Copy the file `/etc/kubernetes/admin.conf` from `ncn-m001` to `~/.kube/config` on the external system.

Reviewer: For consistency with other numbered lists, add an extra space after the number before the text so all lines in a numbered list are indented 4 chars.

```suggestion
1.  Copy the file `/etc/kubernetes/admin.conf` from `ncn-m001` to `~/.kube/config` on the external system.
```

    ```

>     $ mkdir -p ~/.kube

Reviewer: We may want to consider establishing a convention for commands run on external systems like we have for commands run on `ncn-m001` or in the `sat bash` container. Or check if there is such a convention used in other products' docs and use that.

Author: I'm not aware of such a convention already, but I can dig around and try to find out. Otherwise, I think a couple options could be using a hostname like external as the prefix to the prompt, e.g.
```
external $ mkdir -p ~/.kube
```
Though this seems a little clunky to me. I'm open to suggestions otherwise.

Reviewer: I don't feel very strongly about it. Somebody from techpubs might have a preference.

    $ scp ncn-m001:/etc/kubernetes/admin.conf ~/.kube/config
    admin.conf                                       100% 5566   3.0MB/s   00:00
    ```

> 1. Add an new entry for the hostname `kubernetes` to the external system's `/etc/hosts` file.

Reviewer: Indent numbered item text by 4 spaces consistently

> 1. Add an new entry for the hostname `kubernetes` to the external system's `/etc/hosts` file.

Reviewer: ```suggestion
1. Add a new entry for the hostname `kubernetes` to the external system's `/etc/hosts` file.
```

       valid_lft forever preferred_lft forever
   $ IP_ADDRESS=10.102.1.10
   ```

   Once the IP address is determined, add an entry to `/etc/hosts` mapping the IP address to the hostname `kubernetes`.

    ```
    $ echo "${IP_ADDRESS} kubernetes" | sudo tee -a /etc/hosts
    10.102.1.11 kubernetes
    ```

1. Modify `~/.kube/config` to set the cluster server address.

   The value of the `server` key for the `kubernetes` cluster under the `clusters` section should be set to
   `https://kubernetes:6443`.

    ```yaml
    ---
    clusters:
    - cluster:
        certificate-authority-data: REDACTED
        server: https://kubernetes:6443
      name: kubernetes
    ...
    ```

1. Confirm that `kubectl` can access the CSM Kubernetes cluster.

    ```
    $ kubectl get nodes
    NAME       STATUS   ROLES    AGE    VERSION
    ncn-m001   Ready    master   135d   v1.19.9
    ncn-m002   Ready    master   136d   v1.19.9
    ncn-m003   Ready    master   136d   v1.19.9
    ncn-w001   Ready    <none>   136d   v1.19.9
    ncn-w002   Ready    <none>   136d   v1.19.9
    ncn-w003   Ready    <none>   136d   v1.19.9
    ```

1. Use `sat init` to create a configuration file for SAT.

    ```

>     $ sat init -o /home/user/.config/sat/sat.toml

Reviewer: When I was the `root` user on my VM running with the podman wrapper script, I did not need `-o`, and I think it actually would have been wrong for that use case. I think it would also be unnecessary when running in a python virtualenv.

    INFO: Configuration file "/home/user/.config/sat/sat.toml" generated.
    ```

1. Edit the SAT configuration file to set the API and S3 hostnames.

   Externally available API endpoints are given domain names in PowerDNS, so the endpoints in the configuration file
   should each be set to `subdomain.system-name.site-domain`, where `system-name` and `site-domain` are replaced with
   the values specified during `csi config init`, and `subdomain` is the DNS name for the externally available service.

>    More information can be found under the [Externally Exposed
>    Services](https://cray-hpe.github.io/docs-csm/en-12/operations/network/customer_accessible_networks/externally_exposed_services/)
>    page in the CSM documentation.

Reviewer: minor consistency suggestion:

```suggestion
   See [Externally Exposed Services](https://cray-hpe.github.io/docs-csm/en-12/operations/network/customer_accessible_networks/externally_exposed_services/)
   in the CSM documentation for more information.
```

   The API gateway has the subdomain `api`, and S3 has the subdomain `s3`. The S3 endpoint runs on port 8080. The
   following options should be set in the SAT configuration file:

    ```toml

>     ...

Reviewer: These get highlighted in red in the rendered docs, probably because they're not valid toml. Should we just leave them out in these snippets?

Author: I wanted to indicate that there are other lines in the configuration that should not be edited. However, looking at the other parts of the install instructions, such a convention isn't already in use, and we are just putting snippets in verbatim. I don't feel to strongly, so I will just remove the ellipses.

Reviewer: Yeah, I do like to use the ellipsis convention myself, but it just renders a bit funny. Maybe removing the `toml` syntax highlighting would get rid of that. I noticed it didn't happen for yaml syntax.

### Limitations

Most SAT subcommands work by accessing APIs which are reachable via the CAN. However, certain SAT commands depend on
host-based functionality on the management NCNs and will not work from an external system. This includes the following:

- The `platform-services` and `ncn-power` stages of `sat bootsys`
- The local host information displayed by the `--local` option of `sat showrev`

Installing SAT on an external system is not an officially supported configuration. These instructions are provided
"as-is" with the hope that they can useful for users who desire additional flexibility.

Certain additional steps may need to be taken to install and configure SAT depending on the configuration of the
external system in use. These additional steps may include provisioning virtual machines, installing packages, or
configuring TLS certificates, and these steps are outside the scope of this documentation. This section covers only the
steps needed to configure SAT to use externally-accessible API endpoints exposed by CSM.

### Prerequisites

- The external system must be on the Customer Access Network (CAN).
- Python 3.7 or newer is installed on the system.

> - `kubectl`, `openssh`, `git`, and `curl` are installed on the external system.

Author: I was going off of the dependencies we install in the `Dockerfile`, and after seeing this I actually couldn't think of where we use `curl`, but I do actually suggest the user use it to perform stateful resets in `sat swap blade`:
``` % rg curl sat
sat/cli/swap/blade.py
429:                    'curl -k -u root:PASSWORD -X POST -H \\\n'
```
So I guess we have a soft dependency on it, though for anything else they shouldn't need it.

> - `kubectl`, `openssh`, `git`, and `curl` are installed on the external system.

Reviewer 2: Aha, thanks for the pointer. I would not have thought of that usage. I'm good with leaving it.

  documentation for more information.

### Procedure

1.  Create a Python virtual environment.

    ```screen
    $ SAT_VENV_PATH="$(pwd)/venv"
    $ python3 -m venv ${SAT_VENV_PATH}
    $ . ${SAT_VENV_PATH}/bin/activate
    ```

1.  Install the SAT Python package in the virtual environment.

    ```screen
    (venv) $ export PIP_EXTRA_INDEX_URL="https://artifactory.algol60.net/artifactory/csm-python-modules/simple"
    (venv) $ git clone --branch=release/3.17 https://github.com/Cray-HPE/sat.git
    (venv) $ pip install -r sat/requirements.lock.txt
    ...
    (venv) $ pip install ./sat
    ...
    ```

1.  Optional: Add the `sat` virtualenv to the user's `PATH` environment variable.

    If a shell other than `bash` is in use, replace `~/.bash_profile` with the appropriate profile path.

    If the virtualenv is not added to the user's `PATH` environment variable, then
    `source ${SAT_VENV_PATH}/bin/activate` will need to be run before running any SAT commands.

    ```screen
    (venv) $ deactivate
    $ echo export PATH=\"${SAT_VENV_PATH}/bin:${PATH}\" >> ~/.bash_profile
    $ source ~/.bash_profile
    ```

> 1.  Copy the file `/etc/kubernetes/admin.conf` from `ncn-m001` to `~/.kube/config` on the external system.

Reviewer 2: I definitely agree that it would require configuration beyond what is there by default, though I would not be surprised if this is something sites regularly customize. I do feel like doing that configuration is outside the scope of this particular guide, and a site who has set up finer-grained access would probably also be able to set up the client authentication for the external host. I guess we could consider these to be like "reference instructions" that use the default configuration.

I'm having kind of a hard time deciding what to do, but I'm leaning towards just leaving it or at most cautioning that this will give the external host the same level of access to kubernetes as root on the NCNs has.

As an aside I feel like if a site doesn't want to share `admin.conf` and also doesn't know how to set up other user access, then they would just skip these steps and have a partially functional `sat`. But we haven't really tested the different ways in which the non-functional commands would fail.

    link/ether b8:59:9f:1d:d9:0e brd ff:ff:ff:ff:ff:ff
    inet 10.102.1.11/24 brd 10.102.1.255 scope global vlan007
       valid_lft forever preferred_lft forever
    inet6 fe80::ba59:9fff:fe1d:d90e/64 scope link
       valid_lft forever preferred_lft forever
    $ IP_ADDRESS=10.102.1.11
    ```

    On CSM versions prior to 1.2, the CAN IP can be determined by querying the IP address of the `vlan007` interface.

    ```screen
    $ ssh ncn-m001 ip addr show vlan007
    13: vlan007@bond0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether b8:59:9f:1d:d9:0e brd ff:ff:ff:ff:ff:ff
    inet 10.102.1.10/24 brd 10.102.1.255 scope global vlan007
       valid_lft forever preferred_lft forever
    inet6 fe80::ba59:9fff:fe1d:d90e/64 scope link
       valid_lft forever preferred_lft forever
    $ IP_ADDRESS=10.102.1.10
    ```

    Once the IP address is determined, add an entry to `/etc/hosts` mapping the IP address to the hostname `kubernetes`.

    ```screen
    $ echo "${IP_ADDRESS} kubernetes" | sudo tee -a /etc/hosts
    10.102.1.11 kubernetes
    ```

> 1.  Modify `~/.kube/config` to set the cluster server address.

Reviewer 2: Does it work to just put the IP address in without putting an entry in `/etc/hosts`? Like:

```
server: https://10.102.1.11:6443
```

I think that would be better because it saves a step and also doesn't require editing `/etc/hosts`.

Author: Adding the line to `/etc/hosts` is actually required to "fool" the certificate verification. The certificate isn't valid for the CAN IP address on the NCN:
```crayadm@kubetest-final:~> kubectl get nodes
Unable to connect to the server: x509: certificate is valid for 10.16.0.1, 10.252.1.12, 10.252.1.2, 10.252.1.11, not 10.102.1.13
```

Any of the Subject Alternative Names listed for the cert here would work:
```
crayadm@kubetest-final:~> openssl s_client -connect 10.102.1.13:6443 </dev/null 2>/dev/null | openssl x509 -noout -text | grep 'DNS'
                DNS:kubernetes, DNS:kubernetes.default, DNS:kubernetes.default.svc, DNS:kubernetes.default.svc.cluster.local, DNS:ncn-m001, IP Address:10.16.0.1, IP Address:10.252.1.12, IP Address:10.252.1.2, IP Address:10.252.1.11
```
I chose `kubernetes` to add to `/etc/hosts` since it seems most self-explanatory.

Probably a "better" solution would be to add a new SAN to the certificate for the system-specific hostname of `ncn-m001`. This would probably be a CASMPET issue, I can dig into this a bit more. I agree that my solution here is a bit janky, but it seems to work acceptably well.

Reviewer 2: I'm good with it for now, just thinking that editing /etc/hosts is not always something people want to do on their local machines.

Author: Definitely agree on that, but there doesn't seem to be a better solution without having that SAN set up properly on the certificate. I wonder if there is a succinct way of noting this in the ticket. I suppose I could use the above command as a starting point to determine the hostname for the certificate, and then use that in place of `kubernetes` in the hosts file, or otherwise note that they should use `host <ip_address>` to check if there is a hostname associated with the NCN CAN IP address which is also a SAN on the certificate. It does get complicated quickly.
